plugins/cn0540.c

```diff
@@ -0,0 +1,569 @@
+#include <gtk/gtk.h>
+#include <glib.h>
+#include <math.h>
+
+#include "../datatypes.h"
+#include "../osc.h"
+#include "../osc_plugin.h"
+#include "../iio_widget.h"
+#include "../config.h"
+#include "./block_diagram.h"
+
+#define THIS_DRIVER			"CN0540"
+#define ADC_DEVICE			"ad7768-1"
+#define DAC_DEVICE			"ltc2606"
+#define GPIO_CTRL			"one-bit-adc-dac"
+#define VOLTAGE_MONITOR_1		"xadc"
+#define VOLTAGE_MONITOR_2		"ltc2308"
+#define ADC_DEVICE_CH			"voltage0"
+#define DAC_DEVICE_CH			"voltage0"
+#define GPIO_CTRL_FIRST_CH		"voltage0"
+#define VOLTAGE_MONITOR_1_FIRST_CH	"voltage9"
+#define VOLTAGE_MONITOR_2_FIRST_CH	"voltage0"
+
+#define G				0.3
+#define FDA_VOCM_MV			2500
+#define FDA_GAIN			2.667
+#define DAC_BUF_GAIN			1.22
+#define CALIB_MAX_ITER		  	20
+#define XADC_VREF			3.3
+#define XADC_RES			12
+#define NUM_GPIOS			8
+#define NUM_ANALOG_PINS		 	6
+
+struct iio_gpio {
+	struct iio_channel *gpio;
+	char label[30];
+};
+
+struct iio_device *iio_adc;
+struct iio_device *iio_dac;
+struct iio_device *iio_gpio;
+struct iio_device *iio_voltage_mon;
+static struct iio_context *ctx;
+static struct iio_channel *adc_ch;
+static struct iio_channel *dac_ch;
+static struct iio_channel *analog_in[NUM_ANALOG_PINS];
+static struct iio_gpio gpio_ch[NUM_GPIOS];
+static struct iio_widget iio_widgets[25];
+static unsigned int num_widgets;
+
+static GtkWidget *cn0540_panel;
+static GtkCheckButton *tgbtn_shutdown;
+static GtkCheckButton *tgbtn_fda;
+static GtkCheckButton *tgbtn_fda_mode;
+static GtkButton *btn_get_sw_ff;
+static GtkButton *calib_btn;
+static GtkButton *write_btn;
+static GtkButton *read_btn;
+static GtkButton *readvsensor_btn;
+static GtkTextView *sw_ff_status;
+static GtkTextView *shutdown_status;
+static GtkTextView *fda_status;
+static GtkTextView *fda_mode_status;
+static GtkTextView *voltage_status[6];
+static GtkTextView *vshift_log;
+static GtkTextView *vsensor_log;
+static GtkTextView *calib_status;
+static GtkTextBuffer *sw_ff_buffer;
+static GtkTextBuffer *shutdown_buffer;
+static GtkTextBuffer *fda_buffer;
+static GtkTextBuffer *fda_mode_buffer;
+static GtkTextBuffer *voltage_buffer[NUM_ANALOG_PINS];
+static GtkTextBuffer *calib_buffer;
+static GtkTextBuffer *vsensor_buf;
+static GtkTextBuffer *vshift_buf;
+
+static gboolean plugin_detached;
+static gint this_page;
+
+struct osc_plugin plugin;
+
+void delay_ms(int ms)
+{
+	struct timespec ts = { ms / 1000, (ms % 1000) * 1000 * 1000 };
+	nanosleep(&ts, NULL);
+
+}
+
+static gboolean cn0540_get_gpio_state(const char* gpio_name)
+{
+	long long readback = -1;
+
+	for(int idx = 0; idx < NUM_GPIOS; idx++) {
+		if(strstr(gpio_ch[idx].label, gpio_name)) {
+			iio_channel_attr_read_longlong(gpio_ch[idx].gpio, "raw",
+				&readback);
+			break;
+		}
+	}
+
+	if(readback == -1)
+		printf("%s: wrong gpio name: %s\n",__func__, gpio_name);
+
+	return (gboolean)readback;
+}
+
+static void cn0540_set_gpio_state(const char* gpio_name, gboolean state)
+{
+	for(int idx = 0; idx < NUM_GPIOS; idx++) {
```

Comment:
A job fails because of this, check: https://travis-ci.org/github/analogdevicesinc/iio-oscilloscope/jobs/736061729#L4280
Line 103 will cause the same issue, look for all.